src/System.Runtime.Caching/src/System/Runtime/Caching/MemoryCache.cs

```diff
@@ -0,0 +1,904 @@
+                        if (entry.ExpensiveObjectDependency.HasChanged)
+                        {
+                            // If the expensiveObject has been removed explicitly by Cache.Remove,
+                            // return from the SentinelEntry removed callback
+                            // thus effectively removing the SentinelEntry from the cache.
+                            return;
+                        }
+                        break;
+                    case CacheEntryRemovedReason.Evicted:
+                        Dbg.Fail("Reason should never be CacheEntryRemovedReason.Evicted since the entry was inserted as NotRemovable.");
+                        return;
+                    default:
+                        // do nothing if reason is Removed or CacheSpecificEviction
+                        return;
+                }
+
+                // invoke update callback
+                try
+                {
+                    CacheEntryUpdateArguments args = new CacheEntryUpdateArguments(cache, reason, entry.Key, null);
+                    entry.CacheEntryUpdateCallback(args);
+                    Object expensiveObject = (args.UpdatedCacheItem != null) ? args.UpdatedCacheItem.Value : null;
+                    CacheItemPolicy policy = args.UpdatedCacheItemPolicy;
+                    // Only update the "expensive" object if the user returns a new object,
+                    // a policy with update callback, and the change monitors haven't changed.  (Inserting
+                    // with change monitors that have already changed will cause recursion.)
+                    if (expensiveObject != null && IsPolicyValid(policy))
+                    {
+                        cache.Set(entry.Key, expensiveObject, policy);
+                    }
+                    else
+                    {
+                        cache.Remove(entry.Key);
+                    }
+                }
+                catch
+                {
+                    cache.Remove(entry.Key);
+                    // Review: What should we do with this exception?
+                }
+            }
+        }
+
+
+        // private and internal
+
+        internal MemoryCacheStore GetStore(MemoryCacheKey cacheKey)
+        {
+            int hashCode = cacheKey.Hash;
+            if (hashCode < 0)
+            {
+                hashCode = (hashCode == Int32.MinValue) ? 0 : -hashCode;
+            }
+            int idx = hashCode % _storeCount;
+            return _storeRefs[idx].Target;
+        }
+
+        internal object[] AllSRefTargets
+        {
+            get
+            {
+                var allStores = new MemoryCacheStore[_storeCount];
+                for (int i = 0; i < _storeCount; i++)
+                {
+                    allStores[i] = _storeRefs[i].Target;
+                }
+                return allStores;
+            }
+        }
+
+        [SecuritySafeCritical]
```

Comment:
You can remove `[SecuritySafeCritical]` and other security transparency annotations as .NET Core does not support it

Reply:
And all suppressions related to this and code access security eg CA2106 here.